src/route/dataset-route.ts

```diff
@@ -39,72 +48,81 @@ function checkDatasetID(datasetID: string, res: Response): boolean {
+                    {
+                        lang: WELSH,
+                        message: t(translationString, { lng: WELSH })
                     }
+                ],
+                tag: {
+                    name: translationString,
+                    params: {}
                 }
-            ]
-        };
+            }
+        ]
+    };
+}
+
+
+apiRoute.post('/', upload.single('csv'), async (req: Request, res: Response) => {
+    if (!req.file) {
         res.status(400);
-        res.json(err);
+        res.json(errorDtoGenerator('csv', 'errors.no_csv_data'));
         return;
     }
     const lang: string = req.body?.language || req.i18n.language;
     const title: string = req.body?.title;
     if (!title) {
-        const err: ViewErrDTO = {
-            success: false,
-            dataset_id: undefined,
-            errors: [
-                {
-                    field: 'title',
-                    message: [
-                        {
-                            lang: ENGLISH,
-                            message: t('errors.no_title', { lng: ENGLISH })
-                        },
-                        {
-                            lang: WELSH,
-                            message: t('errors.no_title', { lng: WELSH })
-                        }
-                    ],
-                    tag: {
-                        name: 'errors.no_title',
-                        params: {}
-                    }
-                }
-            ]
-        };
         res.status(400);
-        res.json(err);
+        res.json(errorDtoGenerator('title', 'errors.no_title'));
         return;
     }
-    const dataset = Dataset.createDataset(title, 'BetaUser');
-    const saved_dataset_record = await dataset.save();
-    saved_dataset_record.addTitleByString(title, lang);
-    const uploadDTO = await uploadCSVToBlobStorage(req.file?.buffer, saved_dataset_record);
-    if (!uploadDTO.success) {
-        res.status(400);
+    let importRecord: Import;
+    try {
+        importRecord = await uploadCSVToBlobStorage(req.file?.stream, req.file?.mimetype);
+    } catch (err) {
+        logger.error(`An error occured trying to upload the file with the following error: ${e}`);
+        res.status(500);
+        res.json({ message: 'Error uploading file' });
+        return;
     }
+
+    // Everything looks good so far, let's create the dataset and revision records
+    const dataset = new Dataset();
+    dataset.creation_date = new Date();
+    // TODO change how we handle authentication to get the user on the Backend
+    dataset.created_by = 'Test User';
+    const saved_dataset_record = await dataset.save();
+    const datasetInfo = new DatasetInfo();
+    datasetInfo.language = lang;
+    datasetInfo.title = title;
+    datasetInfo.dataset = saved_dataset_record;
+    datasetInfo.save();
+    const revision = new DatasetRevision();
+    revision.dataset = saved_dataset_record;
+    revision.revision_index = 1;
```

Comment:
The data model should have a way of allocating this itself.
It just needs to be monotonically increasing but it doesn't need to be consecutive or even dense. A single global counter will suffice.

Reply:
I thought the idea of this was that only active revisions would have an index.  Abandoned revisions wouldn't.  They'd be around and in theory retrievable but they wouldn't have a place in the index.

Reply:
I think it's just an optimisation for the toposort algorithm so that you don't have to read all revisions and chase all the pointers?
It gives you a correct toposort with a range query and then you can do a single pass to join up the tree.